### PR TITLE
[HigherOrderOp] Should automatically pop modes

### LINKED ISSUE #108282: [HigherOrderOp] Should automatically pop modes
### 🐛 Describe the bug

We introduced some hacks in https://github.com/pytorch/pytorch/pull/107941 to support cond and out_dtype for pre-dispatch tracing. Concretely:
- When we are doing make_fx over a higher order op (HOO), the HOO's [ProxyTorchDispatchMode implementation gets called](https://github.com/pytorch/pytorch/blob/39130c74339ede608a375b94041771468d2fdcdd/torch/_higher_order_ops/cond.py#L263-L264)
- in the current design, every higher order op (HOO) needs to manually handle pre-dispatch in their ProxyTorchDispatchMode implementation. That is, they need to manually pop the pre-dispatch mode (https://github.com/pytorch/pytorch/pull/107941/files#diff-97a6f5bce3f1359473a9ddaf776fdb7b1891dadc70e317660d87f2e2d9cca4eaR141-R154)
- HOO authors also need to manually pop the mode for regular [make_fx tracing](https://github.com/pytorch/pytorch/blob/39130c74339ede608a375b94041771468d2fdcdd/torch/_higher_order_ops/cond.py#L279-L281)
- When regular PyTorch operators encounter a TorchDispatchMode, the PyTorch dispatcher automatically pops the mode before calling the `__torch_dispatch__`.
- Ideally, the HOO dispatch mechanism automatically pops the modes to be consistent with how regular operators work [(e.g. the mode popping should happen here)](https://github.com/pytorch/pytorch/blob/39130c74339ede608a375b94041771468d2fdcdd/torch/_ops.py#L245-L268), so that users do not need to manually pop the modes in their ProxyTorchDispatchMode implementation

To see how dispatch works, put a breakpoint in the [dispatch()](https://github.com/pytorch/pytorch/blob/39130c74339ede608a375b94041771468d2fdcdd/torch/_ops.py#L245-L268) function and try to call `make_fx` over a function that invokes `torch.ops.higher_order.cond`:
```
import torch
import functorch.experimental.control_flow  # for the cond
from functorch import make_fx

def f(x):
   return torch.ops.higher_order.cond(x > 0, torch.sin, torch.cos, [x])

x = torch.tensor(0.5)
gm = make_fx(f, tracing_mode='fake')(x)
```

### Error logs

_No response_

### Minified repro

_No response_

### Versions

main

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #108282

### CODE REVIEW

File: torch/_ops.py [hunk 1]
```diff
@@ -243,19 +244,27 @@ def dispatch(self, dispatch_key, *args, **kwargs):
...
             ), f"Current active mode {curr_mode} not registered"
-            # TODO(voz): The idea behind this is that we do not yet support dispatch by key + mode, only key.
-            return self.python_key_mode_table[type(curr_mode)](*args, **kwargs)
+            handler = self.python_key_mode_table[type(curr_mode)]
+            if isinstance(curr_mode, ProxyTorchDispatchMode):
+                with _pop_mode_temporarily() as mode:
+                    args = (mode,) + args
+                    return handler(*args, **kwargs)
+            else:
+                return handler(*args, **kwargs)
```
Comment (zou3519) [inline torch/_ops.py:263]: We should unconditionally _pop_mode_temporarily() for all modes, not just ProxyTorchDispatchMode. Does this run you into problems?

File: torch/_ops.py [hunk 2]
```diff
@@ -243,19 +244,27 @@ def dispatch(self, dispatch_key, *args, **kwargs):
...
             assert (
                 type(curr_mode) in self.python_key_mode_table
             ), f"Current active mode {curr_mode} not registered"
-            # TODO(voz): The idea behind this is that we do not yet support dispatch by key + mode, only key.
-            return self.python_key_mode_table[type(curr_mode)](*args, **kwargs)
+            handler = self.python_key_mode_table[type(curr_mode)]
+            if isinstance(curr_mode, ProxyTorchDispatchMode):
+                with _pop_mode_temporarily() as mode:
+                    args = (mode,) + args
+                    return handler(*args, **kwargs)
```
Comment (zou3519) [inline torch/_ops.py:261]: nit: handler(mode, *args, **kwargs) should work

File: torch/_ops.py [hunk 3]
```diff
@@ -265,7 +274,17 @@ def dispatch(self, dispatch_key, *args, **kwargs):
...
+                    DispatchKey.PreDispatch, []
+                )
+                handler = self.python_key_mode_table[type(curr_mode)]
+                if len(pre_dispatch_modes) > 0 and isinstance(
+                    curr_mode, ProxyTorchDispatchMode
+                ):
+                    with temporarily_pop_mode(pre_dispatch_modes) as mode:
+                        args = (mode,) + args
+                        return handler(*args, **kwargs)
+                return handler(*args, **kwargs)
```
Comment (zou3519) [inline torch/_ops.py:287]: Same comment as above: we should handle all modes here.

**Review by zou3519 — COMMENTED** [review]
Minor comment about popping the mode no matter the type of mode, but other than that, this looks pretty good

**Review by zou3519 — APPROVED** [review]
Awesome!

**Comment by yanboliang** [comment]
Checked the internal test failures, we need to pop modes for ```executorch``` op ```executorch_call_delegate``` as well. I'll submit a fix to fbcode.

**Pull request closed without merging**